### DISCUSSION TITLE
initrd too large

### DISCUSSION THREAD

**Discussion by Trumeet** [systemd/mkosi#2531] [Q&A]
I'm trying to build a Debian image for an embedded system, and the generated initrd seems to be about 150MiB. It takes too long for the embedded system to copy the whole initrd (inside UKI) from SD card to memory at boot.

Are there any ways to shrink the initrd, or use mkinitcpio?

The only thing (in my scenario) initrd has to do is to mount root in the same disk, and I am seeking ways to shrink the initrd as much as possible. Thanks.

Configuration file:

```
[Distribution]
Distribution=debian
Release=bookworm
Architecture=arm64

[Output]
Format=disk
CompressOutput=false
Output=image
OutputDirectory=out/
CacheDirectory=cache/
BuildDirectory=build/
ImageVersion=1.0
RepartDirectories=repart.d/

[Content]
Bootable=yes
Packages=linux-image-generic
         systemd
         systemd-boot
         systemd-sysv
         udev
         dbus
         openssh-server
         vim
         ncdu
Ssh=yes
RootPassword=redacted
Bootloader=systemd-boot
Locale=en_US.UTF-8
Timezone=America/Vancouver
Hostname=redacted
KernelCommandLine=systemd.volatile=yes loglevel=7 console=ttyS0 root=PARTLABEL=xxxxx

[Validation]
Checksum=yes
SignExpectedPcr=true

[Host]
Incremental=yes
QemuSmp=4
QemuMem=1G
```

With `--debug-workspace`, the `image-initrd.cpio.zst` is 36MiB, which extracts to a 134MiB CPIO. The merged UKI EFI is about 137MiB, and `ukify inspect` shows that the initrd is 105MiB.

**Comment by Trumeet** [comment]
It seems like mkosi is trying to do another brand-new Debian installation for initrd. Are there any ways to use mkinitcpio or another distribution (like Alpine to say?) that offers a smaller image?

**Comment by daandemeyer** [comment]
Grep for settings with KernelModules in them in the docs.

**Reply by Trumeet** [reply]
After inspecting the initrd, it seems like the vast majority of size is occupied by glibc and other binaries, and there isn't much kernel modules inside.

**Reply by daandemeyer** [reply]
In that case you can build your own initrd with mkosi that includes fewer packages. Our default initrds support quite a few different use cases like LVM that would not be required on your embedded system. So you can build your own initrd with mkosi and use it as the initrd of another image using the Initrds= setting. In this initrd you could only install systemd and udev and you'd already end up with a much smaller initrd than the default one.

The other you could do is switch distributions. Me and @keszybz have done quite a bit of work to make sure core Fedora packages are structured in such a way so that they don't pull in unnecessary dependencies. Someone would have to do the same for Debian.

**Reply by Trumeet** [reply]
Thanks. Is it possible for mkosi to create a dedicated initrd using another lighter distribution and use that as the debian initrd? So I don't need to build the initrd myself (which I have to cross compile a tiny rootfs).

**Reply by daandemeyer** [reply]
This is theoretically possible, but I wouldn't recommend it, you're voiding yourself of any support by trying to mix and match distributions like this.

**Comment by daandemeyer** [comment]
To give a bit more context, when we build the initramfs, we don't know which kernel modules you're going to need, so we copy them all into the initramfs. You can customize by using `KernelModulesInitrdInclude=` and `KernelModulesInitrdExclude=`. Note that dracut and mkinitcpio won't know which modules to include either because the image is not intended to be used on the host you build it on. If you build a hostonly initrd with dracut and try to use it on your embedded system there's a high chance it won't boot.
